### PR TITLE
FIX - Improve Link Entry

### LINKED ISSUE #2715: Improve Link Entry
Dialog style input fields open in layers in Gutenberg right now, resulting in visual clutter. Examples of this issue are raised in #2628 and #2204 

For example, when we **insert a Button Block**, we see this:
wordpress - layering 001

Breaking this down, there are 4 elements competing for our attention (numbers 1-4). These elements have a hierarchy in terms of their priority: for example, if the user has inserted a Button Block, they would be most interested in adding a URL and Label for the button first, before tackling alignment, block settings or adding new blocks. Out of the four elements, the link insert has the highest priority:

wordpress - layering 002

We can improve the design to facilitate this.

**Proposed solution: Serialised dialog**
Last year, we developed and user tested a "serialised dialog", which has now been rolled into TinyMCE mobile ([wireframes here](https://invis.io/E7DF1AB62#/252659977_IMG_9737)). In this example, the main toolbar is replaced with a serialised version of the "dialog". The fields in the dialog are ordered by priority, allowing the user to complete the action at any time (without necessarily scrolling through all the input fields). 

We could apply this same concept to input fields/dialogs in Gutenberg: i.e. **when a selection is made that needs extra input (e.g. link), the main block toolbar is replaced by a serialised dialog**. For the button block, this would look something like this

wordpress - layering 003

Thoughts? Improvements?

/cc @karmatosed @melchoyce @mtias

### PR DESCRIPTION
**Pull request by tg-ephox**
## Description
Link entry for Paragraph, Cover Image and Button blocks have all been made similar to the Image block (with the addition of animations on entry/exit)

Formatting toolbars for Cover Image and Image have been modified so that they appear in the single toolbar above the block, so that this is consistent between all blocks, previously these were the only toolbars that were centred.

fixes https://github.com/WordPress/gutenberg/issues/2715

## How Has This Been Tested?

Unit tests, manual browser testing.
## Types of changes

New feature

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style.
- [x] My code follows has proper inline documentation.

### CODE REVIEW

File: blocks/editable/format-toolbar/index.js [hunk 1]
```diff
@@ -40,22 +38,9 @@ const DEFAULT_CONTROLS = [ 'bold', 'italic', 'strikethrough', 'link' ];
 class FormatToolbar extends Component {
```
Comment (youknowriad) [inline blocks/editable/format-toolbar/index.js]: I guess we could refactor this component to be just a functional component?

**Comment by youknowriad** [comment]
screen shot 2017-09-25 at 15 13 05

I'm seeing a light "white" layer on the button border

**Comment by youknowriad** [comment]
I like the fact that we have a unique way of adding links to the components, though, I'm finding that this requries a lot of steps to add the link. 

 - Add a link
 - Tweak the options
 - Validate

Maybe the options could be triggered by a gear button or something. anyway, a more "designy" feedback would be great. @jasmussen @karmatosed

**Comment by youknowriad** [comment]
The button link is not being saved as expected (check the text mode), Not sure if it's related to this PR.

**Comment by youknowriad** [comment]
I noticed that the link popover is not showing up as expected for the inline toolbar (image captions for example)
screen shot 2017-09-25 at 15 20 15

File: blocks/editable/format-toolbar/index.js (hunk 1 above)
Comment (tg-ephox) [reply blocks/editable/format-toolbar/index.js]: I don't think we can as it needs the lifecycle methods?

**Comment by tg-ephox** [comment]
This isn't the correct text for a Button block?
```
<!-- wp:core/button {"align":"center"} -->
<div class="wp-block-button aligncenter"><a href="https://github.com/WordPress/gutenberg"><span>Help build Gutenberg</span></a></div>
<!-- /wp:core/button -->
```

**Comment by youknowriad** [comment]
@tg-ephox for some reason, the `a` tag was empty (no href) for me. I'll dig more later.

**Comment by tg-ephox** [comment]
@youknowriad the dialog that pops in allows you to 'un-link', which will result in a anchor tag with no href, maybe this option shouldn't be available for the Button block?

**Comment by youknowriad** [comment]
Maybe i'm doing something wrong but here's the screencast
kapture 2017-09-27 at 8 49 11

**Comment by tg-ephox** [comment]
@youknowriad I found the URL problem for button

**Comment by tg-ephox** [comment]
hey @annaephox @karmatosed can I pls get some design feedback on this?

**Comment by annaephox** [comment]
Hey @tg-ephox : looks really good to me! 
@karmatosed : our tests from this feature in TinyMCE mobile showed that people find the serialised link dialog intuitive to use. We have tested the Gutenberg version informally at this stage, and results confirm the same

**Comment by youknowriad** [comment]
Great job so far! I'm really liking the consistency.

- Confirmed the button block is working now
- The inline toolbar is placed at the same level of the block toolbar, I'm not sure this was part of the initial plan of this PR, We'll need a design validation for this @mtias @jasmussen @karmatosed 
- Minor: Could we make sure to submit this when we hit "Enter"?
screen shot 2017-10-06 at 09 49 10

File: blocks/library/button/index.js [hunk 2]
```diff
@@ -75,19 +75,9 @@ registerBlockType( 'core/button', {
 					onChange={ ( value ) => setAttributes( { text: value } ) }
 					formattingControls={ [ 'bold', 'italic', 'strikethrough' ] }
 					keepPlaceholderOnFocus
+					extraToolbarButtons={ <UrlInputButton showSettings={ false } url={ url } onChange={ updateUrl } /> }
```
Comment (youknowriad) [inline blocks/library/button/index.js]: Why do we need this prop? Can't we just put this putting in the `BlockControls` component?

File: blocks/url-input/button.js [hunk 3]
```diff
@@ -15,61 +16,184 @@ import { IconButton } from '@wordpress/components';
...
  */
 import UrlInput from './';
+import ToggleControl from '../inspector-controls/toggle-control';
+import { filterURLForDisplay } from '../../editor/utils/url';
+
+const DisplayStep = 'DISPLAY';
+const EditStep = 'EDIT';
+const SettingsStep = 'SETTINGS';
+
+const AllSteps = [ EditStep, SettingsStep, DisplayStep ];
```
Comment (youknowriad) [inline blocks/url-input/button.js]: We have this unofficial style guideline for uppercasing module constants? Could we uppercase those?

`const DISPLAY_STEP`...

File: blocks/url-input/button.js [hunk 4]
```diff
@@ -15,61 +16,184 @@ import { IconButton } from '@wordpress/components';
...
+		} );
+	}
+
+	renderStep( steps ) {
+		const { currentStep, url, isDeleted, opensInNewWindow } = this.state;
+
+		if ( steps[ currentStep ] === EditStep ) {
+			return [
+				<UrlInput key="urlinput" value={ url || '' } onChange={ this.changeLink } required={ ! isDeleted } />,
+				<IconButton key="iconbutton" className="blocks-url-input__unlink" icon="dismiss" label={ __( 'Un-link' ) } disabled={ ! url } onClick={ this.deleteLink } />,
```
Comment (youknowriad) [inline blocks/url-input/button.js]: Could we break this line and the above maybe to improve readability?

File: blocks/url-input/button.js [hunk 5]
```diff
@@ -15,61 +16,184 @@ import { IconButton } from '@wordpress/components';
...
+			}
+
+			return true;
+		} );
+	}
+
+	renderStep( steps ) {
+		const { currentStep, url, isDeleted, opensInNewWindow } = this.state;
+
+		if ( steps[ currentStep ] === EditStep ) {
```
Comment (youknowriad) [inline blocks/url-input/button.js]: Why not a switch?

**Comment by karmatosed** [comment]
@annaephox I am not happy merging it in it's current state. The design discussion is here: https://github.com/WordPress/gutenberg/issues/2715

**Comment by jasmussen** [comment]
Thanks for working on this, sorry for coming in with feedback so late. I'm also coming in fresh, so please let me know if some of the following are known issues, or separate from this PR.

1. The external link toggle can't be set and immediately unset:

extlink

Just click twice on the switch.

2. The link dialog will sometimes move if I click to edit a link, then click around in the dialog a bit:

movingdialog

3. The caret is moved to the start of the input field when I click around. Pretty sure this is the same bug as 2 mentioned above, but perhaps it can help narrow down what happens:

movingcaret

4. The block level link (i.e. for images) looks a bit off. This is what Riad refers to in this comment:

> The inline toolbar is placed at the same level of the block toolbar, I'm not sure this was part of the initial plan of this PR

toolbarposition

Two solutions I can see. The easiest one is probably to let the link dialog box sit stacked below the quick toolbar. 

The other solution is to keep it where it is — i.e cover the quick toolbar. But then we can't have stacked shadows, i.e. the quick toolbar should probably be invisible when the link toolbar is there, and reappear when done. Also, we should make sure they have the same height — right now the link toolbar is slightly taller:

toolbarposition

Given the discussion ongoing in #2983, perhaps it's best to do as little as possible around this area right now.

**Comment by mikekosulin** [comment]
How to setup option: 
Open a new window by default?

**Comment by karmatosed** [comment]
I am closing this as we can always iterate but we reached a point here where we weren't progressing. Let's get feedback in testing and take from there.

**Pull request closed without merging**